entity-framework/core/what-is-new/ef-core-7.0/whatsnew.md

```diff
@@ -3527,3 +3549,783 @@ Notice:
+
+Or, using mapping attributes:
+
+<!--
+    [Index(nameof(Title), Name = "AscendingIndex")]
+    [Index(nameof(Title), Name = "DescendingIndex", AllDescending = true)]
+    public class Post
+    {
+        public int Id { get; set; }
+
+        [MaxLength(64)]
+        public string? Title { get; set; }
+    }
+-->
+[!code-csharp[TwoIndexesByAttribute](../../../../samples/core/Miscellaneous/NewInEFCore7/ModelBuildingSample.cs?name=TwoIndexesByAttribute)]
+
+This generates the following SQL on SQL Server:
+
+```sql
+CREATE INDEX [AscendingIndex] ON [Posts] ([Title]);
+CREATE INDEX [DescendingIndex] ON [Posts] ([Title] DESC);
+```
+
+Indexes over multiple columns can also have different ordering defined for each column. For example:
+
+<!--
+            #region CompositeIndex
+            modelBuilder
+                .Entity<Blog>()
+                .HasIndex(blog => new { blog.Name, blog.Owner })
+                .IsDescending(false, true);
+            #endregion
+-->
+[!code-csharp[CompositeIndex](../../../../samples/core/Miscellaneous/NewInEFCore7/ModelBuildingSample.cs?name=CompositeIndex)]
+
+Or, using a mapping attribute:
+
+<!--
+    [Index(nameof(Name), nameof(Owner), IsDescending = new[] { false, true })]
+    public class Blog
+    {
+        public int Id { get; set; }
+
+        [MaxLength(64)]
+        public string? Name { get; set; }
+
+        [MaxLength(64)]
+        public string? Owner { get; set; }
+
+        public List<Post> Posts { get; } = new();
+    }
+-->
+[!code-csharp[CompositeIndexByAttribute](../../../../samples/core/Miscellaneous/NewInEFCore7/ModelBuildingSample.cs?name=CompositeIndexByAttribute)]
+
+This results in the following SQL when using SQL Server:
+
+```sql
+CREATE INDEX [IX_Blogs_Name_Owner] ON [Blogs] ([Name], [Owner] DESC);
+```
+
+### Mapping attribute for composite keys
+
+EF7 introduces a new mapping attribute (aka "data annotation") for specifying the primary key property or properties of any entity type. Unlike <xref:System.ComponentModel.DataAnnotations.KeyAttribute?displayProperty=nameWithType>, [PrimaryKeyAttribute](https://github.com/dotnet/efcore/blob/main/src/EFCore.Abstractions/PrimaryKeyAttribute.cs) is placed on the entity type class rather than on the key property. For example:
+
+```csharp
+    [PrimaryKey(nameof(PostKey))]
+    public class Post
+    {
+        public int PostKey { get; set; }
+    }
+```
+
+This makes it a natural fit for defining composite keys:
+
+<!--
+    [PrimaryKey(nameof(PostId), nameof(CommentId))]
+    public class Comment
+    {
+        public int PostId { get; set; }
+        public int CommentId { get; set; }
+    }
+-->
+[!code-csharp[CompositePrimaryKey](../../../../samples/core/Miscellaneous/NewInEFCore7/ModelBuildingSample.cs?name=CompositePrimaryKey)]
+
+Defining the index on the class also means it can be used to specify private properties or fields as keys, even though these would usually be ignored when building the EF model. For example:
```

Comment:
Huh, didn't know we didn't support [Key] on private properties/fields... Any specific reason for that?

Reply:
We do...but only if the member is already mapped. It won't cause it to be mapped.

Reply:
Should we consider enabling this (i.e. mapping the member because it has [Key] on it)? Or would that potentially map something unwanted (because [Key] is being used for some other non-EF framework)?

Reply:
Yes, and yes. :-D I made a note to file an issue; it's something I have been thinking about for a while, and something that Diego had ideas on.